runtime/jvm/src/main/kotlin/kotlinx/serialization/Resolving.kt

```diff
@@ -67,7 +68,19 @@ fun serializerByTypeToken(type: Type): KSerializer<Any> = when(type) {
     }
     is ParameterizedType -> {
         val rootClass = (type.rawType as Class<*>)
-        val args = (type.actualTypeArguments)
+        val args = (type.actualTypeArguments).map { argument ->
+            when (argument) {
+                is WildcardType -> {
+                    argument.upperBounds.first().let { firstUpperBound ->
+                        when(firstUpperBound) {
+                            is Class<*> -> firstUpperBound
```

Comment:
Well, same question. Is it always `Class<*>`? Would this work with `List<List<List<...>>>`, etc? 

I can suggest you an approach to handle WildCard types as a separate clause, not inside ParameterizedType clause. This probably will look like this: `is ParameterizedType -> serializerByTypeToken(type.upperBounds.first())`. Probably with some additional checks around.

Reply:
that ended up being the smarter choice indeed
well i reordered it a bit, factored unwrapping of WildcardType into a private extension val
it now supports nested generic containers (list, map), nested arrays and the mix of both
the code to support GenericArray ended up being a bit too ugly for my taste, thanks to java's types